File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListActivity.java
package com.kjjejones42.audiobookplayer.display;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.ProgressDialog;
import android.content.ComponentName;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SearchView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.work.OneTimeWorkRequest;
import androidx.work.Operation;
import androidx.work.WorkManager;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.kjjejones42.audiobookplayer.AudioBook;
import com.kjjejones42.audiobookplayer.R;
import com.kjjejones42.audiobookplayer.Utils;
import com.kjjejones42.audiobookplayer.database.AudiobookDatabase;
import com.kjjejones42.audiobookplayer.player.MediaPlaybackService;
import com.kjjejones42.audiobookplayer.player.PlayActivity;

import java.util.List;

public class DisplayListActivity extends AppCompatActivity {

    private static final String INTENT_UPDATE_MODEL = "com.kjjejones42.audiobookplayer.UPDATE";
    public static final String INTENT_PLAY_FILE = "com.kjjejones42.audiobookplayer.PLAY";
    public static final String INTENT_START_PLAYBACK = "com.kjjejones42.audiobookplayer.start";

    private final String[] PERMISSIONS = new String[] {
            Manifest.permission.READ_MEDIA_IMAGES,
            Manifest.permission.READ_MEDIA_AUDIO
    };

    private static final int MY_PERMISSIONS_REQUEST_READ_STORAGE = 3;
    private DisplayListViewModel model;
    private DisplayListAdapter mAdapter;
    private RecyclerView recyclerView;
    private TextView emptyView;
    private SearchView searchView;
    private MediaControllerCompat controller;
    private MediaBrowserCompat browser;

    private final ActivityResultLauncher<String[]> activityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.RequestMultiplePermissions(),
            x -> askUserForDirectory()
    );

    public void chooseDirectory(@SuppressWarnings("unused") MenuItem item) {
        activityResultLauncher.launch(PERMISSIONS);
    }

    private void askUserForDirectory() {
        final String message = "Loading. Please wait...";
        OneTimeWorkRequest request = new OneTimeWorkRequest.Builder(FileScannerWorker.class).build();
        final ProgressDialog dialog = ProgressDialog.show(this, "", message, true);
        WorkManager.getInstance(this).enqueue(request).getState().observe(this, workInfo -> {
            if (!workInfo.getClass().equals(Operation.State.IN_PROGRESS.class)) {
                try {
                    Intent intent = new Intent(getApplicationContext(), DisplayListActivity.class);
                    intent.putExtra(INTENT_UPDATE_MODEL, true);
                    dialog.cancel();
                    startActivity(intent);
                } catch (Exception e) {
                    Utils.logError(e, this);
                    e.printStackTrace();
                }
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == MY_PERMISSIONS_REQUEST_READ_STORAGE) {
            int total = grantResults.length;
            for (int i = 0; i < permissions.length; i++) {
                total += grantResults[i];
            }
            boolean allApproved = total == grantResults.length;
            String result = allApproved ? "All" : "Not all";
            Toast.makeText(this, result + " permissions granted.", Toast.LENGTH_LONG).show();
            if (allApproved) {
                askUserForDirectory();
            }
        }
    }

    @Override
    public void onBackPressed() {
        if (searchView != null && !searchView.isIconified()) {
            searchView.setQuery("",false);
            searchView.clearFocus();
            searchView.setIconified(true);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu, menu);
        searchView = (SearchView) menu.findItem(R.id.app_bar_search).getActionView();
        assert searchView != null;
        searchView.setSubmitButtonEnabled(false);
        searchView.setOnCloseListener(() -> {
            mAdapter.filter(null);
            return false;
        });
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                mAdapter.filter(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                mAdapter.filter(newText);
                return false;
            }
        });
        return true;
    }

    @SuppressLint("Range")
    void updateScreen() {
        List<AudioBook> list = model.getSavedBooks(this).getValue();
        if (list != null && list.isEmpty()) {
            recyclerView.setVisibility(View.GONE);
            emptyView.setVisibility(View.VISIBLE);
        } else {
            recyclerView.setVisibility(View.VISIBLE);
            emptyView.setVisibility(View.GONE);
        }

    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (intent.getBooleanExtra(INTENT_UPDATE_MODEL, false)) {
            model.loadFromDatabase(this);
        }
        updateScreen();
    }

    private void resumeMostRecentBook() {
        AudioBook mostRecent = AudiobookDatabase.getInstance(this).audiobookDao().getMostRecentBook();
        if (mostRecent != null && mostRecent.getLastSavedTimestamp() > 0) {
            Intent intent = new Intent(this, MediaPlaybackService.class);
            intent.putExtra(PlayActivity.INTENT_AUDIOBOOK, mostRecent.displayName);
            intent.putExtra(PlayActivity.INTENT_INDEX, mostRecent.getPositionInTrackList());
            startService(intent);
        }
    }


    public void onFloatingActionButtonClick(@SuppressWarnings("unused") View v) {
        if (controller.getPlaybackState() == null || controller.getPlaybackState().getState() != PlaybackStateCompat.STATE_PLAYING) {
            resumeMostRecentBook();
        } else {
            controller.getTransportControls().pause();
        }
    }

    final MediaBrowserCompat.ConnectionCallback connectionCallbacks = new MediaBrowserCompat.ConnectionCallback(){
        @Override
        public void onConnected() {
            super.onConnected();
            controller = new MediaControllerCompat(
                    DisplayListActivity.this,
                    browser.getSessionToken());
            controller.registerCallback(new MediaControllerCompat.Callback() {
                @Override
                public void onPlaybackStateChanged(PlaybackStateCompat state) {
                    if (state != null) {
                        FloatingActionButton fab = findViewById(R.id.fab);
                        int icon = (state.getState() == PlaybackStateCompat.STATE_PLAYING) ?
                                R.drawable.ic_pause : R.drawable.ic_play;
                        fab.setImageDrawable(AppCompatResources.getDrawable(getBaseContext(), icon));
                    }
                    super.onPlaybackStateChanged(state);
                }
            });
        }
    } ;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_display_list);

        recyclerView = findViewById(R.id.recyclerView);
        emptyView = findViewById(R.id.empty_view);

        model = new ViewModelProvider(this).get(DisplayListViewModel.class);
        browser = new MediaBrowserCompat(this, new ComponentName(this, MediaPlaybackService.class), connectionCallbacks, null);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new DisplayListAdapter(model, recyclerView, this);
        recyclerView.setAdapter(mAdapter);
        model.loadFromDatabase(this);
        updateScreen();
    }

    @Override
    protected void onResume() {
        super.onResume();
        model.loadFromDatabase(this);
        updateScreen();
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (browser != null) {
            browser.connect();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (browser != null) {
            browser.disconnect();
        }
    }


}

File: app/src/main/java/com/kjjejones42/audiobookplayer/display/DisplayListActivity.kt
//package com.kjjejones42.audiobookplayer.display
//
//import android.Manifest
//import android.annotation.SuppressLint
//import android.content.ComponentName
//import android.content.Intent
//import android.os.Bundle
//import android.support.v4.media.MediaBrowserCompat
//import android.support.v4.media.session.MediaControllerCompat
//import android.support.v4.media.session.PlaybackStateCompat
//import android.util.TypedValue
//import android.view.Menu
//import android.view.MenuItem
//import android.view.View
//import android.view.WindowInsets
//import android.widget.SearchView
//import android.widget.Toast
//import androidx.activity.result.contract.ActivityResultContracts.RequestMultiplePermissions
//import androidx.appcompat.app.AppCompatActivity
//import androidx.appcompat.content.res.AppCompatResources
//import androidx.core.view.ViewCompat
//import androidx.core.view.WindowCompat
//import androidx.core.view.WindowInsetsCompat
//import androidx.lifecycle.ViewModelProvider
//import androidx.recyclerview.widget.LinearLayoutManager
//import androidx.work.OneTimeWorkRequest
//import androidx.work.WorkManager
//import com.kjjejones42.audiobookplayer.AudioBook
//import com.kjjejones42.audiobookplayer.R
//import com.kjjejones42.audiobookplayer.database.AudiobookDatabase.Companion.getInstance
//import com.kjjejones42.audiobookplayer.databinding.ActivityDisplayListBinding
//import com.kjjejones42.audiobookplayer.player.MediaPlaybackService
//import com.kjjejones42.audiobookplayer.player.PlayActivity.Companion.INTENT_AUDIOBOOK
//import com.kjjejones42.audiobookplayer.player.PlayActivity.Companion.INTENT_INDEX
//
//class DisplayListActivity : AppCompatActivity() {
//
//    companion object {
//        private val PERMISSIONS = arrayOf(
//            Manifest.permission.READ_MEDIA_IMAGES,
//            Manifest.permission.READ_MEDIA_AUDIO
//        )
//        const val INTENT_PLAY_FILE: String = "com.kjjejones42.audiobookplayer.PLAY"
//        const val INTENT_START_PLAYBACK: String = "com.kjjejones42.audiobookplayer.start"
//
//        private const val MY_PERMISSIONS_REQUEST_READ_STORAGE = 3
//    }
//
//    private lateinit var mAdapter: DisplayListAdapter
//    private lateinit var viewbinding: ActivityDisplayListBinding
//    private var searchView: SearchView? = null
//    private var controller: MediaControllerCompat? = null
//    private lateinit var browser: MediaBrowserCompat
//
//    private val activityResultLauncher = registerForActivityResult(RequestMultiplePermissions()) { askUserForDirectory() }
//
//    fun chooseDirectory(item: MenuItem?) {
//        activityResultLauncher.launch(PERMISSIONS)
//    }
//
//    private fun askUserForDirectory() {
////        val message = "Loading. Please wait..."
//        val request = OneTimeWorkRequest.Builder(FileScannerWorker::class.java).build()
////        val dialog = ProgressDialog.show(this, "", message, true)
//        WorkManager.getInstance(this)
//            .enqueue(request)
////            .state
////            .observe(this) { if (it.javaClass != IN_PROGRESS::class.java) dialog.cancel() }
//    }
//
//    override fun onRequestPermissionsResult(
//        requestCode: Int,
//        permissions: Array<String>,
//        grantResults: IntArray
//    ) {
//        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
//        if (requestCode == MY_PERMISSIONS_REQUEST_READ_STORAGE) {
//            val allApproved = grantResults.all { it == 1 }
//            val result = if (allApproved) "All" else "Not all"
//            Toast.makeText(this, "$result permissions granted.", Toast.LENGTH_LONG).show()
//            if (allApproved) {
//                askUserForDirectory()
//            }
//        }
//    }
//
//    override fun onBackPressed() {
//        searchView?.takeIf { !it.isIconified }?.apply {
//            setQuery("", false)
//            clearFocus()
//            isIconified = true
//        } ?: super.onBackPressed()
//    }
//
//    override fun onCreateOptionsMenu(menu: Menu): Boolean {
//        menuInflater.inflate(R.menu.menu, menu)
//        searchView = menu.findItem(R.id.app_bar_search).actionView as SearchView?
//        searchView?.apply {
//            isSubmitButtonEnabled = false
//            setOnCloseListener {
//                mAdapter.filter(null)
//                false
//            }
//            setOnQueryTextListener(object : SearchView.OnQueryTextListener {
//                override fun onQueryTextSubmit(query: String): Boolean {
//                    mAdapter.filter(query)
//                    return false
//                }
//
//                override fun onQueryTextChange(newText: String): Boolean {
//                    mAdapter.filter(newText)
//                    return false
//                }
//            })
//        }
//        return true
//    }
//
//    override fun onCreate(savedInstanceState: Bundle?) {
//        setTheme(R.style.AppTheme)
//        super.onCreate(savedInstanceState)
//        WindowCompat.setDecorFitsSystemWindows(window, false)
//
//        viewbinding = ActivityDisplayListBinding.inflate(layoutInflater)
//        setContentView(viewbinding.root)
//
//        ViewCompat.setOnApplyWindowInsetsListener(viewbinding.root) { view, insets ->
//            val systemBarsInsets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
//
//            // 2. Apply Insets to Child Views
//            // We adjust the view's padding or margins based on the insets.
//
//            // Example: Applying top inset as padding to the root view's content
//            view.setPadding(
//                view.paddingLeft,
//                systemBarsInsets.top, // Apply top inset here
//                view.paddingRight,
//                view.paddingBottom
//            )
//            insets
//        }
//
//        val model = ViewModelProvider(this)[DisplayListViewModel::class.java]
//        val serviceComponent = ComponentName(this, MediaPlaybackService::class.java)
//        browser = MediaBrowserCompat(this, serviceComponent, connectionCallbacks, null)
//
//        viewbinding.recyclerView.setLayoutManager(LinearLayoutManager(this))
//        mAdapter = DisplayListAdapter(model, viewbinding.recyclerView, this)
//        viewbinding.recyclerView.setAdapter(mAdapter)
//
//        getInstance(this).audiobookDao().allAndObserve.observe(this) { model.setBooks(it) }
//        model.savedBooks.observe(this) { this.updateScreen(it) }
//
//        askUserForDirectory()
//    }
//
//    private fun resumeMostRecentBook() {
//        getInstance(this).audiobookDao().mostRecentBook
//            ?.takeIf { it.lastSavedTimestamp > 0 }
//            ?.let {
//                val intent = Intent(this, MediaPlaybackService::class.java)
//                intent.putExtra(INTENT_AUDIOBOOK, it.displayName)
//                intent.putExtra(INTENT_INDEX, it.positionInTrackList)
//                startService(intent)
//            }
//    }
//
//    fun onFloatingActionButtonClick(@Suppress("unused") v: View?) {
//        controller?.apply {
//            if (playbackState == null || playbackState.state != PlaybackStateCompat.STATE_PLAYING) {
//                resumeMostRecentBook()
//            } else {
//                transportControls.pause()
//            }
//        }
//    }
//
//    private val connectionCallbacks: MediaBrowserCompat.ConnectionCallback =
//        object : MediaBrowserCompat.ConnectionCallback() {
//            override fun onConnected() {
//                super.onConnected()
//                controller = MediaControllerCompat(this@DisplayListActivity, browser.sessionToken)
//                controller!!.registerCallback(object : MediaControllerCompat.Callback() {
//                    override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
//                        val icon = if (state.state == PlaybackStateCompat.STATE_PLAYING) R.drawable.ic_pause else R.drawable.ic_play
//                        viewbinding.fab.setImageDrawable(AppCompatResources.getDrawable(baseContext, icon))
//                        super.onPlaybackStateChanged(state)
//                    }
//                })
//            }
//        }
//
//    @SuppressLint("Range")
//    fun updateScreen(list: List<AudioBook?>?) {
//        val listVisible = list != null && list.isEmpty()
//        viewbinding.recyclerView.visibility = if (listVisible) View.GONE else View.VISIBLE
//        viewbinding.emptyView.visibility = if (listVisible) View.VISIBLE else View.GONE
//    }
//
//
//    override fun onStart() {
//        super.onStart()
//        browser.connect()
//    }
//
//    override fun onStop() {
//        super.onStop()
//        browser.disconnect()
//    }
//}
//
